### ISSUE TITLE
Notification with Error Code are not supported

### ISSUE THREAD

**Issue by sbernard31** [eclipse-leshan/leshan#1560]
Playing **leshan-server-demo**, I faced : 

```
[WARN] Unable to handle notification of [Hu6ujaOuLP:/3442/0/1110/0]  
org.eclipse.leshan.core.request.exception.InvalidResponseException: Unexpected response code [4.04] for SingleObservation [path=/3442/0/1110/0, id=Ox656143955aec5b23, contentFormat=SENML_JSON(110), registrationId=Hu6ujaOuLP, context={}]
	at org.eclipse.leshan.server.californium.endpoint.ServerCoapMessageTranslator.createObservation(ServerCoapMessageTranslator.java:92)
	at org.eclipse.leshan.server.californium.endpoint.CaliforniumServerEndpointsProvider$2.onNotification(CaliforniumServerEndpointsProvider.java:168)
	at org.eclipse.californium.core.network.CoapEndpoint$NotificationDispatcher.onNotification(CoapEndpoint.java:811)
	at org.eclipse.californium.core.network.BaseMatcher$2.onResponse(BaseMatcher.java:250)
	at org.eclipse.californium.core.coap.Request.setResponse(Request.java:916)
	at org.eclipse.californium.core.server.ServerMessageDeliverer.deliverResponse(ServerMessageDeliverer.java:280)
	at org.eclipse.californium.core.network.stack.BaseCoapStack$StackTopAdapter.receiveResponse(BaseCoapStack.java:214)
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:89)
	at org.eclipse.californium.oscore.ObjectSecurityContextLayer.receiveResponse(ObjectSecurityContextLayer.java:274)
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:89)
	at org.eclipse.californium.core.network.stack.ExchangeCleanupLayer.receiveResponse(ExchangeCleanupLayer.java:105)
	at org.eclipse.californium.core.network.stack.ObserveLayer.receiveResponse(ObserveLayer.java:154)
	at org.eclipse.californium.core.network.stack.BlockwiseLayer.receiveResponse(BlockwiseLayer.java:862)
	at org.eclipse.californium.core.network.stack.ReliabilityLayer.receiveResponse(ReliabilityLayer.java:308)
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:89)
	at org.eclipse.californium.oscore.ObjectSecurityLayer.receiveResponse(ObjectSecurityLayer.java:394)
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveResponse(AbstractLayer.java:89)
	at org.eclipse.californium.core.network.stack.BaseCoapStack.receiveResponse(BaseCoapStack.java:132)
	at org.eclipse.californium.core.network.CoapEndpoint$1.receiveResponse(CoapEndpoint.java:324)
	at org.eclipse.californium.core.network.UdpMatcher$4.run(UdpMatcher.java:457)
	at org.eclipse.californium.elements.util.SerialExecutor$1.run(SerialExecutor.java:292)
	at org.eclipse.californium.core.network.CoapEndpoint$6.run(CoapEndpoint.java:1367)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

This happens when an observed node is removed. In that case a last NOT_FOUND notification should be sent.

See [RFC7641 -  Observing Resources in the Constrained Application Protocol (CoAP) § 3.2.  Notifications](https://www.rfc-editor.org/rfc/rfc7641#section-3.2) says : 

>  In the event that the resource changes in a way that would cause a
   normal GET request at that time to return a non-2.xx response (for
   example, when the resource is deleted), the server sends a
   notification with an appropriate response code (such as 4.04 Not
   Found) and removes the client's entry from the list of observers of
   the resource.  Non-2.xx responses do not include an Observe Option.

**Comment by sbernard31** [comment]
<details>
  <summary>I put this code crappy code that I could need again when I will work on **java-coap** provider</summary>

```java
server.getObservationService().addListener(new ObservationListener() {

    @Override
    public void onResponse(CompositeObservation observation, Registration registration,
            ObserveCompositeResponse response) {
        System.out.println("==================");
        System.out.println("RESPONSE " + response);
        System.out.println("==================");
    }

    @Override
    public void onResponse(SingleObservation observation, Registration registration, ObserveResponse response) {
        System.out.println("==================");
        System.out.println("SINGLE RESPONSE " + response);
        System.out.println("==================");

    }

    @Override
    public void onError(Observation observation, Registration registration, Exception error) {
        System.out.println("==================");
        System.out.println("ERROR " + error);
        error.printStackTrace();
        System.out.println("==================");

    }

    @Override
    public void newObservation(Observation observation, Registration registration) {
        System.out.println("==================");
        System.out.println("NEW OBS " + observation.getId());
        System.out.println("==================");

    }

    @Override
    public void cancelled(Observation observation) {
        System.out.println("==================");
        System.out.println("CANCELLED " + observation.getId());
        System.out.println("==================");

    }
});
```

</details>

**Comment by sbernard31** [comment]
Issue was closed automatically because I integrated #1561 in `master` but this only fixes the Californium provider.

Bug for **java-coap** is still there so I reopen.
